gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -81,26 +86,42 @@ public BatcherImpl(
       BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor,
       UnaryCallable<RequestT, ResponseT> unaryCallable,
       RequestT prototype,
-      BatchingSettings batchingSettings) {
+      BatchingSettings batchingSettings,
+      ScheduledExecutorService executor) {
     this.batchingDescriptor =
         Preconditions.checkNotNull(batchingDescriptor, "batching descriptor cannot be null");
     this.unaryCallable = Preconditions.checkNotNull(unaryCallable, "callable cannot be null");
     this.prototype = Preconditions.checkNotNull(prototype, "request prototype cannot be null");
     this.batchingSettings =
         Preconditions.checkNotNull(batchingSettings, "batching setting cannot be null");
+
+    Preconditions.checkNotNull(executor, "executor cannot be null");
+    final Runnable pushCurrentBatchRunnable =
+        new Runnable() {
+          @Override
+          public void run() {
+            sendBatch();
+          }
+        };
+    long delay = batchingSettings.getDelayThreshold().toMillis();
+    scheduledFuture =
+        executor.scheduleWithFixedDelay(
+            pushCurrentBatchRunnable, delay, delay, TimeUnit.MILLISECONDS);
```

Comment:
I think the Runnable needs to have a WeakReference to the Batcher

Reply:
Please have a look at `PushCurrentOpenBatch`, I think this would unschedule the future from GCed Batcher.

Reply:
Ok, lets move conversation to that class